File: Difficulty: Medium/Game with String/game-with-string.cpp
class Solution {
  public:
    int minValue(string& s, int k) {
        vector<int> freq(26, 0); // Frequency array for 26 lowercase letters

        // Count frequencies
        for (char ch : s) {
            freq[ch - 'a']++;
        }

        // Remove k characters greedily from the highest frequency char
        while (k--) {
            sort(freq.begin(), freq.end()); // Sort to bring max freq to end
            if (freq[25] == 0) {            // If all characters are removed
                break;
            }
            freq[25]--; // Remove one occurrence from highest freq char
        }

        int result = 0;
        for (int val : freq) {
            result += val * val; // Sum of squares of remaining frequencies
        }

        return result;
    }
};

File: Difficulty: Medium/Game with String/game-with-string.java
class Solution {
    public int minValue(String s, int k) {
        // Step 1: Count frequency of each character and get maximum frequency
        int[] alphaFreq = new int[26];
        int maxi = 0;

        for (char ch : s.toCharArray()) {
            alphaFreq[ch - 'a']++;
            maxi = Math.max(maxi, alphaFreq[ch - 'a']);
        }

        // Step 2: Build a bucket array where index = frequency and value = count of chars with that freq
        int[] freq = new int[maxi + 1];
        for (int i = 0; i < 26; i++) {
            if (alphaFreq[i] > 0) {
                freq[alphaFreq[i]]++;
            }
        }

        int maxFreq = maxi;

        // Step 3: Reduce highest frequencies greedily
        while (k > 0 && maxFreq > 0) {
            int countAtMax = freq[maxFreq];

            if (countAtMax <= k) {
                k -= countAtMax;
                freq[maxFreq - 1] += countAtMax;
                freq[maxFreq]      = 0;
                maxFreq--;
            } else {
                freq[maxFreq]     -= k;
                freq[maxFreq - 1] += k;
                k = 0;
            }
        }

        // Step 4: Calculate sum of squares of remaining frequencies
        int result = 0;
        for (int i = 1; i <= maxi; i++) {
            result += i * i * freq[i];
        }

        return result;
    }
}